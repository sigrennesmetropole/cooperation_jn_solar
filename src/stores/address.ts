import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'

export const useAddressStore = defineStore('address', () => {
  const latitude: Ref<number> = ref(0)
  const longitude: Ref<number> = ref(0)
  const address: Ref<string> = ref('')
  const screenshotAddress: Ref<Blob | null> = ref(null)

  function setLatitudeAndLongitude(
    new_latitude: number,
    new_longitude: number
  ) {
    latitude.value = new_latitude
    longitude.value = new_longitude
  }

  function setAddress(new_address: string) {
    address.value = new_address
  }

  function resetLatitureAndLongitude() {
    setLatitudeAndLongitude(0, 0)
  }

  function resetAddress() {
    setAddress('')
  }

  return {
    latitude,
    longitude,
    setLatitudeAndLongitude,
    address,
    setAddress,
    resetAddress,
    resetLatitureAndLongitude,
    screenshotAddress,
  }
})
