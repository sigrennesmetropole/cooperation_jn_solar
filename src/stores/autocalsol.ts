import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'
import type { AutocalsolResult } from '@/model/autocalsol.model'

export const useAutocalsolStore = defineStore('autocalsolStore', () => {
  const autocalsolResult: Ref<AutocalsolResult | null> = ref(null)

  function setAutocalsolResult(result: AutocalsolResult | null) {
    autocalsolResult.value = result
  }

  function resetAutocalsolResult() {
    autocalsolResult.value = null
  }

  return {
    autocalsolResult,
    setAutocalsolResult,
    resetAutocalsolResult,
  }
})
