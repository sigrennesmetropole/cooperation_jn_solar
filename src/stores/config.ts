import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'
import type { ConfigType } from '@sigrennesmetropole/cooperation_jn_common_ui'

export const useConfigStore = defineStore('configStore', () => {
  const config: Ref<ConfigType | null> = ref(null)

  function setConfig(newConfig: ConfigType | null) {
    config.value = newConfig
  }

  return {
    config,
    setConfig,
  }
})
