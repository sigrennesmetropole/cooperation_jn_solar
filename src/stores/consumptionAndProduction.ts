import type { Ref } from 'vue'
import { ref } from 'vue'
import { defineStore } from 'pinia'

export const useConsumptionAndProductionStore = defineStore(
  'consumptionAndProduction',
  () => {
    const annualConsumption: Ref<number> = ref(0)

    function setAnnualConsumption(new_annual_consumption: number) {
      annualConsumption.value = new_annual_consumption
    }

    function resetAnnualConsumption() {
      setAnnualConsumption(0)
    }

    return {
      annualConsumption,
      setAnnualConsumption,
      resetAnnualConsumption,
    }
  }
)
