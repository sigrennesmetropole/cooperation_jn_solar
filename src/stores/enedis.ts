import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'

export const useEnedisStore = defineStore('enedisStore', () => {
  const isEnedisRedirection: Ref<boolean> = ref(false)

  function setIsEnedisRedirection(newValue: boolean) {
    isEnedisRedirection.value = newValue
  }

  return {
    isEnedisRedirection,
    setIsEnedisRedirection,
  }
})
