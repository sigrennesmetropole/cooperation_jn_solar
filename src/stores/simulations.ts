import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'
import router from '@/router'
import { getNumberFromConfig } from '@/services/configService'

export function getDefaultConsumption() {
  return getNumberFromConfig('consumption.default_consumption')
}

export const useSimulationStore = defineStore('simulation', () => {
  const currentStep: Ref<number> = ref(1)
  const currentSubStep: Ref<number> = ref(1)

  const availableSteps = [
    { step: 1, subStep: 1, isFinal: false }, // Choose roof side
    { step: 2, subStep: 1, isFinal: false }, // Select obstacle
    { step: 2, subStep: 2, isFinal: false }, // Select the number of solar panel
    { step: 2, subStep: 3, isFinal: false }, // No roof panel available
    { step: 3, subStep: 1, isFinal: false }, // Energy saving information
    { step: 3, subStep: 2, isFinal: false }, // Select input invoice or connect to Linky
    { step: 3, subStep: 3, isFinal: true }, // Input tarif/invoice
    { step: 3, subStep: 4, isFinal: true }, // Connect to Linky meter
  ]

  const restartEndSimulation: Ref<boolean> = ref(false)

  function setCurrentStep(step: number) {
    currentStep.value = step
  }

  function setCurrentSubStep(subStep: number) {
    currentSubStep.value = subStep
  }

  function indexOfCurrentStepAndSubStep() {
    return availableSteps.findIndex((availableStep) => {
      return (
        availableStep.step == currentStep.value &&
        availableStep.subStep == currentSubStep.value
      )
    })
  }

  function isCurrentStepFinal() {
    const currentIndex = indexOfCurrentStepAndSubStep()
    return availableSteps[currentIndex].isFinal
  }

  function getIndexPreviousStepNotFinal(currentIndex: number) {
    for (let index = currentIndex - 1; index >= 0; index--) {
      if (!availableSteps[index].isFinal) {
        return index
      }
    }
    return null
  }

  function goToFinalView() {
    router.push('end-simulation')
  }

  function goToPreviousStep() {
    const currentIndex = indexOfCurrentStepAndSubStep()
    if (currentIndex - 1 < 0) {
      return
    }
    const indexPreviousStepNotFinal = getIndexPreviousStepNotFinal(currentIndex)
    if (indexPreviousStepNotFinal !== null) {
      setCurrentStep(availableSteps[indexPreviousStepNotFinal].step)
      setCurrentSubStep(availableSteps[indexPreviousStepNotFinal].subStep)
    }
    if (currentStep.value == 2 && currentSubStep.value == 3) {
      goToPreviousStep()
    }
  }

  function goToNextStep() {
    const currentIndex = indexOfCurrentStepAndSubStep()
    if (availableSteps[currentIndex].isFinal) {
      goToFinalView()
      return
    }
    if (currentIndex + 1 >= availableSteps.length) {
      return
    }
    setCurrentStep(availableSteps[currentIndex + 1].step)
    setCurrentSubStep(availableSteps[currentIndex + 1].subStep)
    if (currentStep.value == 2 && currentSubStep.value == 3) {
      goToNextStep()
    }
  }

  function resetSimulation() {
    setCurrentStep(1)
    setCurrentSubStep(1)
  }

  function shouldShowSolarPanelLayer(): boolean {
    // NOTE(IS): If you want to debug visually the location of solar panel,
    // you set solar panel is shown in the Select obstacle step (step: 2, subStep: 1)
    // Then use back button after generating the solar panel
    if (currentStep.value === 3) {
      return true
    }
    if (currentStep.value === 2 && currentSubStep.value === 2) {
      return true
    }
    return false
  }

  function isSelectObstacleStep() {
    return currentStep.value === 2 && currentSubStep.value === 1
  }

  return {
    currentStep,
    currentSubStep,
    restartEndSimulation,
    setCurrentStep,
    setCurrentSubStep,
    goToPreviousStep,
    goToNextStep,
    isCurrentStepFinal,
    goToFinalView,
    resetSimulation,
    shouldShowSolarPanelLayer,
    isSelectObstacleStep,
  }
})
