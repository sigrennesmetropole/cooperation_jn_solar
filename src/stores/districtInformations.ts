import { ref } from 'vue'
import type { Ref } from 'vue'
import { defineStore } from 'pinia'
import type { Viewpoint } from '@vcmap/core'
import type { Feature } from 'ol'
import { useInteractionsStore } from '@/stores/interactions'
import SelectDistrictInteraction from '@/interactions/selectDistrictInteractions'
import { RENNES_LAYER, useLayersStore } from '@/stores/layers'

export const useDistrictStore = defineStore('districtInformations', () => {
  const interactionsStore = useInteractionsStore()
  const layerStore = useLayersStore()

  const districtCode: Ref<number> = ref(0)
  const districtName: Ref<string> = ref('')
  const districtConsumption: Ref<number> = ref(0)
  const districtProduction: Ref<number> = ref(0)
  const districtNumberInstallations: Ref<number> = ref(0)
  const districtPointFeature: Ref<Feature | null> = ref(null)
  const previousViewPoint: Ref<Viewpoint | null> = ref(null)
  const newPointAbscissa: Ref<number> = ref(0)
  const newPointOrdinate: Ref<number> = ref(0)
  const checkboxChecked: Ref<boolean> = ref(false)
  const canBeDisplayed: Ref<boolean> = ref(true)
  function setDistrictIrisCode(newDistrictCode: number) {
    districtCode.value = newDistrictCode
  }

  function setDistrictName(newDistrictName: string) {
    districtName.value = newDistrictName
  }

  function setDistrictConsumption(newDistrictConsumption: number) {
    districtConsumption.value = newDistrictConsumption
  }

  function setDistrictProduction(newDistrictProduction: number) {
    districtProduction.value = newDistrictProduction
  }

  function setDistrictNumberInstallations(
    newDistrictNumberInstallations: number
  ) {
    districtNumberInstallations.value = newDistrictNumberInstallations
  }

  function setDistrictInformations(
    newDistrictCode: number,
    newDistrictName: string,
    newDistrictConsumption: number,
    newDistrictProduction: number,
    newDistrictNumberInstallations: number
  ) {
    setDistrictIrisCode(newDistrictCode)
    setDistrictName(newDistrictName)
    setDistrictConsumption(newDistrictConsumption)
    setDistrictProduction(newDistrictProduction)
    setDistrictNumberInstallations(newDistrictNumberInstallations)
  }

  function setNewPointFeatureOnSelectedDistrict(newPointFeature: Feature) {
    districtPointFeature.value = newPointFeature
  }

  function setNewCoordinates(newAbscissa: number, newOrdinate: number) {
    newPointAbscissa.value = newAbscissa
    newPointOrdinate.value = newOrdinate
  }

  function setCheckboxChecked(newAction: boolean) {
    checkboxChecked.value = newAction
    if (newAction && canBeDisplayed) {
      layerStore.enableLayer(RENNES_LAYER.iris)
      interactionsStore.enableInteraction(SelectDistrictInteraction)
    } else {
      layerStore.disableLayer(RENNES_LAYER.iris)
      interactionsStore.disableInteraction(SelectDistrictInteraction)
    }
  }

  function setCanBeDisplayed(isFarEnough: boolean) {
    canBeDisplayed.value = isFarEnough
    if (!isFarEnough) {
      setCheckboxChecked(false)
    }
  }

  function resetDistrictStore() {
    setDistrictInformations(0, '', 0, 0, 0)
  }

  return {
    districtCode,
    districtName,
    districtConsumption,
    districtProduction,
    districtNumberInstallations,
    districtPointFeature,
    previousViewPoint,
    newPointAbscissa,
    newPointOrdinate,
    checkboxChecked,
    canBeDisplayed,
    setDistrictIrisCode,
    setDistrictName,
    setDistrictConsumption,
    setDistrictProduction,
    setDistrictNumberInstallations,
    setDistrictInformations,
    setNewPointFeatureOnSelectedDistrict,
    setNewCoordinates,
    setCheckboxChecked,
    setCanBeDisplayed,
    resetDistrictStore,
  }
})
