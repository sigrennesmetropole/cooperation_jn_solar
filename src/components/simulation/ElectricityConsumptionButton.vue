<script setup lang="ts">
import rectangle from '@/assets/icons/rectangle.svg'
import expandArrow from '@/assets/icons/expand-small-bigger-retract-smaller-big.svg'
import { useSimulationStore, DEFAULT_CONSUMPTION } from '@/stores/simulations'
import { useConsumptionAndProductionStore } from '@/stores/consumptionAndProduction'

const simulationStore = useSimulationStore()
const consumptionAndProductionStore = useConsumptionAndProductionStore()

function skipStep() {
  consumptionAndProductionStore.setAnnualConsumption(DEFAULT_CONSUMPTION)
  simulationStore.goToFinalView()
}
</script>

<template>
  <div class="flex flex-col gap-2.5 py-6 font-dm-sans">
    <button
      class="bg-black shadow-sm rounded-lg gap-3 px-4 py-3 items-center flex flex-row justify-center"
      @click="simulationStore.currentSubStep = 3"
      id="button-manual-input"
    >
      <span class="text-white text-base font-medium"
        >Saisir les informations de ma facture</span
      >
    </button>
    <div class="flex flex-row justify-between gap-[22px]">
      <img :src="rectangle" />
      <p class="text-sm font-dm-sans font-normal text-neutral-400 my-4">OU</p>
      <img :src="rectangle" />
    </div>
    <button
      class="border border-black rounded-lg gap-3 px-4 py-3 mb-6 items-center flex flex-row justify-center"
      @click="simulationStore.currentSubStep = 4"
      id="button-linky"
    >
      <img :src="expandArrow" class="h-5 w-5" />
      <span class="text-black text-base font-medium"
        >Connecter mon compteur Linky</span
      >
    </button>
    <p class="text-xs font-dm-sans font-normal text-neutral-600">
      Ma consommation n'est traitée que le temps de la simulation et n' est pas
      conservée ultérieurement. Je peux également obtenir une simulation
      partielle sans renseigner ma consommation.
      <span
        id="skipText"
        class="font-medium underline decoration-1 cursor-pointer"
        @click="skipStep()"
        >Passer cette étape</span
      >
    </p>
  </div>
</template>
