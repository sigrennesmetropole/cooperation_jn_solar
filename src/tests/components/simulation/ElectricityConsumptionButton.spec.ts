import ElectricityConsumptionButton from '@/components/simulation/ElectricityConsumptionButton.vue'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { useConsumptionAndProductionStore } from '@/stores/consumptionAndProduction'
import { createTestingPinia } from '@pinia/testing'
import { DEFAULT_CONSUMPTION } from '@/stores/simulations'

describe('ElectricityConsumptionButton.vue', () => {
  let wrapper: VueWrapper

  beforeEach(async () => {
    wrapper = mount(ElectricityConsumptionButton, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            stubActions: false,
            stubPatch: false,
            fakeApp: true,
          }),
        ],
      },
    })
  })

  it('emits "clickAnnualConsumption" with "manual" when "Saisir les informations de ma facture" button is clicked', async () => {
    const manualInputButton = wrapper.find('#button-manual-input')
    await manualInputButton.trigger('click')

    expect(wrapper.emitted('clickAnnualConsumption')).toBeTruthy()
    expect(wrapper.emitted('clickAnnualConsumption')![0]).toEqual(['manual'])
  })

  it('emits "clickAnnualConsumption" with "linky" when "Connecter mon compteur Linky" button is clicked', async () => {
    const linkyButton = wrapper.find('#button-linky')
    await linkyButton.trigger('click')

    expect(wrapper.emitted('clickAnnualConsumption')).toBeTruthy()
    expect(wrapper.emitted('clickAnnualConsumption')![0]).toEqual(['linky'])
  })
  it('update consumption information when "Passer cette Ã©tape" is clicked', async () => {
    const consumptionAndProductionStore = useConsumptionAndProductionStore()

    const skipText = wrapper.find('[id="skipText"]')
    await skipText.trigger('click')
    expect(consumptionAndProductionStore.annualConsumption).toBe(
      DEFAULT_CONSUMPTION
    )
  })
})
