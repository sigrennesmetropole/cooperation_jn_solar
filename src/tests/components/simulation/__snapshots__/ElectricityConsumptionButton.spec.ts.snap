// Vitest Snapshot v1

exports[`ElectricityConsumptionButton.vue > renders the component correctly 1`] = `
"<div class=\\"flex flex-col gap-2.5 py-6 font-dm-sans\\"><button class=\\"bg-black shadow-sm rounded-lg gap-3 px-4 py-3 items-center flex flex-row justify-center\\" id=\\"button-manual-input\\"><span class=\\"text-white text-base font-medium\\">Saisir les informations de ma facture</span></button>
  <div class=\\"flex flex-row justify-between gap-[22px]\\"><img src=\\"/src/assets/icons/rectangle.svg\\">
    <p class=\\"text-sm font-dm-sans font-normal text-neutral-400 my-4\\">OU</p><img src=\\"/src/assets/icons/rectangle.svg\\">
  </div><button class=\\"border border-black rounded-lg gap-3 px-4 py-3 mb-6 items-center flex flex-row justify-center\\" id=\\"button-linky\\"><img src=\\"/src/assets/icons/expand-small-bigger-retract-smaller-big.svg\\" class=\\"h-5 w-5\\"><span class=\\"text-black text-base font-medium\\">Connecter mon compteur Linky</span></button>
  <p class=\\"text-xs font-dm-sans font-normal text-neutral-600\\"> Ma consommation n'est traitée que le temps de la simulation et n' est pas conservée ultérieurement. Je peux également obtenir une simulation partielle sans renseigner ma consommation. <span id=\\"skipText\\" class=\\"font-medium underline decoration-1 cursor-pointer\\">Passer cette étape</span></p>
</div>"
`;
